pkg/pillar/cmd/loguploader/loguploader.go

```diff
@@ -756,9 +776,28 @@ func sendToCloud(ctx *loguploaderContext, data []byte, iter int, fName string, f
 			log.Tracef("sendToCloud: sent ok, file time %v, latency %d, content %s",
 				filetime, latency, string(contents))
 		} else {
-			if resp.StatusCode == http.StatusServiceUnavailable {
+			if resp.StatusCode == http.StatusServiceUnavailable { // status code 503
 				serviceUnavailable = true
 			} else if isResp4xx(resp.StatusCode) {
+				// status code 429
+				if resp.StatusCode == http.StatusTooManyRequests {
+					ctx.metrics.LastTooManyReqTime = time.Now()
+					ctx.metrics.NumTooManyRequest++
+					backoffMode = true
+					if ctx.backoffExprTimer != nil {
+						ctx.backoffExprTimer.Stop()
+					}
+					ctx.backoffExprTimer = time.NewTimer(backoffMaxTimeout * time.Second)
+					// double the upload interval up to a max
+					currentIntv := ctx.metrics.CurrUploadIntvSec * 2
```

Comment:
Do you want to double it for every request which gets a 429, as opposed to only double once per some window/time period?
Asked differently,  once you enter into backoffMode==true do you want to double for every additional 429 you get and wait for the full hour until any reduction?

TCP is multipicative decrease (in rate) and additive increase, checked per CWND or per WND. That would correspond to multiplying the CurrUploadIntvSec when you get 429 and doing some reduction to CurrUploadIntvSec when you get a 200 response.

Reply:
I think you are right. maybe too much to do doubling. I'll modify

Reply:
Doubling once per some epoch is fine if there was a 429 in that epoch. Question is whether that epoch should be each attempt to send, and that would be fine if you only try to send once if you get an error. But I don't know what happens if you have N files waiting and you get some error (429 or other) after some of them - will you continue trying to send all N or not? That affects whether the epoch should be each send or each sequence of N sends.

Comment 2 is that if you don't see a 429 (during an epoch), then it would make sense to decrease CurrUploadIntvSec by a bit.

Reply:
This change does not change the original scheme, which is for any N files in directory, we always try to send the earliest in timestamp first. So if we are not successfully sending a file, regardless of witch error, we'll try to send the same file in the next try(or move it to failedToSend directory); and when to send next one is determined either by the 'normal scheduling' (basically file number based upload interval) or this patch of 429 enabled scheme.
I would like to keep the interval the same during the 429-enabled time , the current 429 rate-limit is basically per device based from cloudflare, so lower that would cross the triggering point again for sure. Until we have per cluster based rate-limit, may make sense to lower that.